db/error_handler.cc

```diff
@@ -465,42 +456,36 @@ const Status& ErrorHandler::SetBGError(const Status& bg_status,
       // this case
       // TODO:  a better way to set or clean the retryable IO error which
       // happens during compaction SST file write.
-      if (bg_error_stats_ != nullptr) {
-        RecordTick(bg_error_stats_.get(), ERROR_HANDLER_AUTORESUME_COUNT);
-      }
+      RecordStats({ERROR_HANDLER_AUTORESUME_COUNT}, {} /* int_histograms */);
       ROCKS_LOG_INFO(
           db_options_.info_log,
           "ErrorHandler: Compaction will schedule by itself to resume\n");
       // Not used in this code path.
       new_bg_io_err.PermitUncheckedError();
       return bg_error_;
-    } else if (BackgroundErrorReason::kFlushNoWAL == reason ||
-               BackgroundErrorReason::kManifestWriteNoWAL == reason) {
+    }
+
+    Status::Severity severity;
+    if (BackgroundErrorReason::kFlushNoWAL == reason ||
+        BackgroundErrorReason::kManifestWriteNoWAL == reason) {
       // When the BG Retryable IO error reason is flush without WAL,
       // We map it to a soft error. At the same time, all the background work
       // should be stopped except the BG work from recovery. Therefore, we
       // set the soft_error_no_bg_work_ to true. At the same time, since DB
       // continues to receive writes when BG error is soft error, to avoid
       // to many small memtable being generated during auto resume, the flush
       // reason is set to kErrorRecoveryRetryFlush.
-      Status bg_err(new_bg_io_err, Status::Severity::kSoftError);
-      CheckAndSetRecoveryAndBGError(bg_err);
+      severity = Status::Severity::kSoftError;
       soft_error_no_bg_work_ = true;
       context.flush_reason = FlushReason::kErrorRecoveryRetryFlush;
-      recover_context_ = context;
-      return StartRecoverFromRetryableBGIOError(bg_io_err);
     } else {
-      Status bg_err(new_bg_io_err, Status::Severity::kHardError);
-      CheckAndSetRecoveryAndBGError(bg_err);
-      recover_context_ = context;
-      return StartRecoverFromRetryableBGIOError(bg_io_err);
+      severity = Status::Severity::kHardError;
     }
+    Status bg_err(new_bg_io_err, severity);
+    CheckAndSetRecoveryAndBGError(bg_err);
+    recover_context_ = context;
+    return StartRecoverFromRetryableBGIOError(bg_io_err);
   } else {
-    if (bg_error_stats_ != nullptr) {
```

Comment:
It seems to me if this stats recording happens, below block in `HandleKnownErrors` will happen too:
```
  if (bg_error_stats_ != nullptr) {
    RecordTick(bg_error_stats_.get(), ERROR_HANDLER_BG_ERROR_COUNT);
    RecordTick(bg_error_stats_.get(), ERROR_HANDLER_BG_ERROR_COUNT_MISSPELLED);
  }
```

And since this is the only caller of `HandleKnownErrors`, I removed its stats recording, and combined it into this block. This resulted in the common stats recording block at the beginning of this function. Let me know if this doesn't look alright.